pkg/api/steve/supportconfigs/generator.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/rancher/rancher/pkg/types/config"
+	"github.com/sirupsen/logrus"
+	authzv1 "k8s.io/api/authorization/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	authv1 "k8s.io/client-go/kubernetes/typed/authorization/v1"
+)
+
+const (
+	// Endpoint The endpoint that this URL is accessible at - used for routing and for SARs
+	Endpoint            = "/v1/generateSUSERancherSupportConfig"
+	cspAdapterNamespace = "cattle-csp-system"
+	cspAdapterConfigmap = "csp-config"
+	tarContentType      = "application/x-tar"
+	logPrefix           = "support-config-generator"
+)
+
+type GeneratorHandler struct {
+	ConfigMaps           v1.ConfigMapInterface
+	SubjectAccessReviews authv1.SubjectAccessReviewInterface
+	ConfigHandler        supportConfigHandler
+}
+
+func NewGeneratorHandler(scaledContext *config.ScaledContext) GeneratorHandler {
+	return GeneratorHandler{
+		ConfigMaps:           scaledContext.Core.ConfigMaps(cspAdapterNamespace),
+		SubjectAccessReviews: scaledContext.K8sClient.AuthorizationV1().SubjectAccessReviews(),
+		ConfigHandler:        supportConfigHandler{},
+	}
+}
+
+func (h *GeneratorHandler) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	authorized, err := h.authorize(request)
+	if err != nil {
+		_, _ = writer.Write([]byte("forbidden"))
+		writer.WriteHeader(http.StatusForbidden)
+		logrus.Errorf("[%s] Failed to authorize user with error: %s", logPrefix, err.Error())
+		return
+	}
+	if !authorized {
+		_, _ = writer.Write([]byte("forbidden"))
+		writer.WriteHeader(http.StatusForbidden)
+		return
+	}
+	logrus.Infof("[%s] Generating supportconfig", logPrefix)
+	archive, err := h.generateSupportConfig()
+	logrus.Infof("[%s] Done Generating supportconfig", logPrefix)
+	if err != nil {
+		logrus.Errorf("[%s] Error when generating support config %v", logPrefix, err)
+		_, _ = writer.Write([]byte("Internal Server Error"))
+		writer.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	writer.Header().Set("Content-Type", tarContentType)
+	writer.Header().Set("Content-Disposition", "attachment; filename=\"supportConfig.tar\"")
+	n, err := io.Copy(writer, archive)
+	logrus.Debugf("[%s] wrote %v bytes in archive response", logPrefix, n)
+}
+
+// authorize checks to see if the user can get the non-resource url of Endpoint. If not, returns an error, otherwise nil
+func (h *GeneratorHandler) authorize(r *http.Request) (bool, error) {
+	userInfo, ok := request.UserFrom(r.Context())
+	if !ok {
+		return false, fmt.Errorf("unable to extract user info from context")
+	}
+	response, err := h.SubjectAccessReviews.Create(r.Context(), &authzv1.SubjectAccessReview{
+		Spec: authzv1.SubjectAccessReviewSpec{
+			// This auth method prevents collision with future k8s resources and allows this to be added to custom rules
+			NonResourceAttributes: &authzv1.NonResourceAttributes{
+				Verb: "get",
+				Path: Endpoint,
```

Comment:
To summarize, this endpoint isn't set or given anywhere, the idea is that no one should have access to this endpoint besides admin, is that right ?

Reply:
Correct. The reason that I did it this way is technically you could use K8s rbac to add this to another role if you really wanted to. We don't, but a customer could.